include/aws/io/stream.h

```diff
@@ -0,0 +1,108 @@
+#ifndef AWS_IO_STREAM_H
+#define AWS_IO_STREAM_H
+
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/io/io.h>
+
+struct aws_input_stream;
+struct aws_byte_buf;
+
+/*
+ * For seek calls, where in the stream to seek from.
+ * CUR support can come later
+ * Intentionally mirror libc constants
+ */
+enum aws_stream_seek_basis { AWS_SSB_BEGIN = 0, AWS_SSB_END = 2 };
+
+struct aws_stream_status {
+    bool is_end_of_stream;
+    bool is_valid;
+};
+
+typedef int(
+    aws_input_stream_seek_fn)(struct aws_input_stream *stream, aws_off_t offset, enum aws_stream_seek_basis basis);
+typedef int(aws_input_stream_read_fn)(struct aws_input_stream *stream, struct aws_byte_buf *dest, size_t *amount_read);
+typedef int(aws_input_stream_get_status_fn)(struct aws_input_stream *stream, struct aws_stream_status *status);
+typedef int(aws_input_stream_get_length_fn)(struct aws_input_stream *stream, size_t *out_length);
```

Comment:
I don't feel strongly about this, just making sure we've thought about it:

Is there any compelling reason to make length part of aws_stream_status, instead of a separate query?

Reply:
length may not be a valid query so it would also need an indicator variable.  Additionally, our notion of length is a relatively constant property but the other status members are more dynamic.  So I think it's ok to be separate or eventually folded into a separate structure.